[JENKINS-43444] Connect new JNLP Agent should be one-click

batmat opened the pull request:
https://issues.jenkins-ci.org/browse/JENKINS-43444
Not only provide the `java -jar ...` command, but also a `curl` or Windows command to download the 
then run the slave.jar. Plus a nowadays more and more common "Copy" button to save yet a bit more time.

Screenshot of the slightly changed UI (if you click on the `Linux` button, which is the default display when the page shows up):

![image](https://cloud.githubusercontent.com/assets/223853/24820838/75f2faf2-1beb-11e7-86cf-32b650db6810.png)


If you click on the `Windows` button:

![image](https://cloud.githubusercontent.com/assets/223853/24820814/5b5aea60-1beb-11e7-9697-2e6bb130da61.png)

From a technical standpoint, essentially:
* defines 'copiable' class that then enables the auto addition of a copy
  button for marked textareas
* Use of `data-windows` and `data-linux` attributes to store command for respective
  platforms. I could have used a JS var to fill the field when showing the
  page, but I wanted to have something that would still show up in case JS
  is disabled or at least a bit broken (and not an empty textarea in that
  case)

batmat committed:
[JENKINS-43444] agent startup in one click

Simplify a bit the start of a JNLP agent

Not only provide the `java -jar ...` command, but also the `curl` or
Win command to download and start the slave.jar.

Essentially:
* defines 'copiable' class that then enables the auto addition of a copy
  button for marked textareas
* Use of data-windows and data-linux to store command for respective
  platforms. I could have used a JS var a fill the field when showing the
  page, but I wanted to have something that would still show up in case JS
  is disabled or at least a bit broken (and not an empty textarea in that
  case)

oleg-nenashev commented:
Maybe `curl ... && java ...` then? Just to make it copy&paster

batmat commented:
@oleg-nenashev why not, but actually it already works by copy pasting. It will just then execute indeed the first line right away, do you think this is an issue? Maybe so.

daniel-beck commented:
`&&` is invalid in `fish` and possibly other shells, so we're just going to get complaints from super special snowflakes.

daniel-beck commented:
This ignores Windows and will just make @jtnord cry.

batmat commented:
I actually chose curl instead of what for that reason : I think it's more common nowadays to have curl in general (than wget), even on Windows. But now I'm at it, I'm thinking about making something similar to what GitHub has on some articles : the command could be switched with a click + a copy to buffer button.
Also: my thought was also that people on even more special platform would just copy the second line as now.

If James thinks there's a better command for Windows though I can consider using it.

oleg-nenashev commented:
Actually curl is also not guaranteed to be present on the Unix systems. So maybe you could add a comment like "or just download it using the hyperlink below" after the command sample

jtnord commented:
Windows would use  powershell. 
`Invoke-WebRequest -Uri $theurl -OutFile slave.jar`

Requires IE so won't work apparently on server core installations.

oleg-nenashev commented:
So do we want to merge it as is? @batmat I am waiting for the feedback from you regarding CURL

batmat commented:
Yeah, right. Neither is `java` guaranteed to be present actually ;-). I'm marking this as WIP as I need more time to think about what I could do to make it still usable for the core use case I wanted to achieve (make a common Linux agent JNLP Startup quicker thank right-click, type wget or curl, paste the link, back to jenkins, copy/paste the java -jar ... command)

batmat commented:
@daniel-beck @oleg-nenashev @jtnord I've played a bit with JS, and here's a screenshot of what I have right now:

![image](https://cloud.githubusercontent.com/assets/223853/24772777/e92bff84-1b12-11e7-9c24-ba3fb0a96d25.png)

I put the existing text in a readonly textarea. The copy button directly copies the buffer, like it's becoming more common nowadays. WDYT?

Now, looking into making it switchable easily for Windows agents.

oleg-nenashev commented:
@batmat Looks good to me. Maybe it would be preferable to have "copy" icons instead of buttons, similarly to GitHub and other such services. Maybe frontend guys could suggest something, CC @kzantow , @recena, @tfennelly

batmat commented:
Yeah, it was a quick first attempt and I didn't want to lose time looking for icons/glyphs. I had indeed in mind to make it a bit more fancy when other things are OK.

batmat commented:
@oleg-nenashev @daniel-beck @jtnord WDYT, I updated the PR code + the screenshots above (https://github.com/jenkinsci/jenkins/pull/2775#issue-211816469).

I would like to avoid spending more time on this PR to add an icon or so immediately. 

This looks quite fine to me right now from a usability perspective, and at least I think it does already improve the previous state.

If you think this is acceptable as is, I would like to merge it as is, and consider filing a followup PR to replace the "Copy" button with something like a copy icon, but would prefer not to do right now.

WDYT?

oleg-nenashev commented on core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly:
```diff
@@ -66,10 +117,16 @@ THE SOFTWARE.
             </j:otherwise>
           </j:choose>
+          <button id="linux"   style="width:10em" onclick="switchText('textarea.copiable[data-linux]','data-linux');  ">Linux</button>
+          <button id="windows" style="width:10em" onclick="switchText('textarea.copiable[data-linux]','data-windows');">Windows</button>
```
`textarea.copiable[data-linux]` is suspicious

batmat replied:
Actually, I understand your suspicion, but it's not. The right way to do the selection I had in mind would have been something like `textarea.copiable[data-linux&&data-windows]` but I considered it overkill.

oleg-nenashev commented on core/src/main/resources/hudson/slaves/JNLPLauncher/main.jelly:
```diff
@@ -51,13 +98,17 @@ THE SOFTWARE.
-                <pre>java${it.launcher.vmargs == null ? '' : ' ' + it.launcher.vmargs} -jar <a href="${rootURL}/jnlpJars/slave.jar">slave.jar</a> -jnlpUrl ${h.inferHudsonURL(request)}${it.url}slave-agent.jnlp</pre>
+                  <textarea class="copiable" readonly="readonly"
+                            data-windows="Invoke-WebRequest -Uri $theurl -OutFile slave.jar\njava${it.launcher.vmargs == null ? '' : ' ' + it.launcher.vmargs} -jar slave.jar -jnlpUrl ${h.inferHudsonURL(request)}${it.url}slave-agent.jnlp"
+                            data-linux="curl -O ${h.inferHudsonURL(request)}jnlpJars/slave.jar\njava${it.launcher.vmargs == null ? '' : ' ' + it.launcher.vmargs} -jar slave.jar -jnlpUrl ${h.inferHudsonURL(request)}${it.url}slave-agent.jnlp">curl -O ${h.inferHudsonURL(request)}jnlpJars/slave.jar
```
Not every Linux contains CURL, And not Every non-Windows system is Linux

oleg-nenashev replied:
My recommendation would be to select a neutral term like POSIX or Unix-alike

batmat replied:
Yeah, I agree that from a technical standpoint, it's true. But I think putting POSIX instead of Linux would be less clear for a majority of users. Maybe Unix, yes, but well, it's probably also a minority of agents (pretty sure stats would back that btw).

oleg-nenashev replied:
what would you think about just replacing the text by `CURL`? Seems to be pretty straightforward and there is no "Linux" mention which may piss off other distribution maintainers.

batmat replied:
Sorry, missed your message @oleg-nenashev :disappointed: 
Yeah, I think this may make sense.

But then I leave Windows for the other? Makes sense. Doing it.

batmat replied:
Done on the button/visible text only. But well, now I think I probably should do that too even on the code. Maybe.

recena commented:
I will try to review the code related with the UI.

recena commented:
@batman I can do the PR followup for polishing the UI/UX. If you agree.

batmat commented:
@recena that would be awesome! Team work :-). :handshake: :raised_hands:

batmat committed:
Use cURL instead of Linux to be more specific

cURL may be absent from Linux, be present on Windows, and so on.
Better be specific to try and clarify.

oleg-nenashev reviewed (changes requested):
Sorry for another 🐛 . "Invoke-WebRequest" is not available in Batch. It is specific to PowerShell only. So I would change it as well

oleg-nenashev commented:
Maybe a default "With downloaded jar" option is also helpful.

batmat commented:
Right, will do that later today or during the WE hopefully. NIT: I think you meant :-1: instead of :bug: @oleg-nenashev ;-)

oleg-nenashev commented:
@batmat Sorry, the bug emoji is so cute in Windows, cannot stop using it. 

<img width="619" alt="capture" src="https://cloud.githubusercontent.com/assets/3000480/25285939/4c6b6940-26bc-11e7-8235-7de835cf4969.PNG">

daniel-beck commented:
Would be preferable if somewhere on the page there still remains a regular download link. `curl` is only an improvement if it doesn't replace a generic way to download the file.

oleg-nenashev commented:
@daniel-beck do you require the change here?

daniel-beck commented:
@oleg-nenashev Yes, since it got removed from https://github.com/jenkinsci/jenkins/pull/2775/files#diff-63a487bb5047df29065dc7ddbb149176L60

oleg-nenashev commented:
There will be a conflict with #2875 since I move the code to the plugin

daniel-beck commented:
Does this need escaping of weird agent names? See e.g. https://issues.jenkins-ci.org/browse/JENKINS-26658 for someone who didn't expect to have to escape `%` in batch.

daniel-beck commented:
Provided feedback outside the formal review mechanism, so unrequested myself.

batmat commented:
Closing for now as I don't find time to finish it. Might reopen some day.
